src/common/profile.c

```diff
@@ -0,0 +1,350 @@
+	}
+	return chunk;
+}
+
+static struct daos_profile *
+profile_alloc(int op_cnt)
+{
+	struct daos_profile *dp;
+
+	D_ALLOC_PTR(dp);
+	if (dp == NULL)
+		return NULL;
+
+	D_ALLOC_ARRAY(dp->dp_ops, op_cnt);
+	if (dp->dp_ops == NULL) {
+		D_FREE_PTR(dp);
+		return NULL;
+	}
+	dp->dp_ops_cnt = op_cnt;
+
+	return dp;
+}
+
+void
+daos_profile_destroy(struct daos_profile *dp)
+{
+	struct daos_profile_chunk *dpc;
+	struct daos_profile_chunk *tmp;
+	int			  i;
+
+	for (i = 0; i < dp->dp_ops_cnt; i++) {
+		struct daos_profile_op	*dpo;
+
+		dpo = &dp->dp_ops[i];
+
+		d_list_for_each_entry_safe(dpc, tmp, &dpo->dpo_chunk_list,
+					   dpc_chunk_list)
+			profile_chunk_destroy(dpc);
+
+		d_list_for_each_entry_safe(dpc, tmp, &dpo->dpo_chunk_idle_list,
+					   dpc_chunk_list)
+			profile_chunk_destroy(dpc);
+	}
+
+	if (dp->dp_dir_path)
+		D_FREE(dp->dp_dir_path);
+
+	D_FREE_PTR(dp);
+}
+
+static int
+profile_get_new_chunk(struct daos_profile_op *dpo)
+{
+	struct daos_profile_chunk *chunk;
+
+	if (!d_list_empty(&dpo->dpo_chunk_idle_list)) {
+		/* Try to get one from idle list */
+		chunk = d_list_entry(dpo->dpo_chunk_idle_list.next,
+				     struct daos_profile_chunk,
+				     dpc_chunk_list);
+		d_list_move_tail(&chunk->dpc_chunk_list, &dpo->dpo_chunk_list);
+		dpo->dpo_chunk_cnt++;
+		D_ASSERT(chunk->dpc_chunk_offset == 0);
+		D_ASSERT(dpo->dpo_chunk_cnt <= dpo->dpo_chunk_total_cnt);
+	} else if (dpo->dpo_chunk_total_cnt < DEFAULT_CHUNK_CNT_LIMIT) {
+		/* Allocate new one */
+		chunk = profile_chunk_alloc(DEFAULT_CHUNK_SIZE);
+		if (chunk == NULL)
+			return -DER_NOMEM;
+
+		d_list_add_tail(&chunk->dpc_chunk_list,
+				&dpo->dpo_chunk_list);
+		dpo->dpo_chunk_total_cnt++;
+		dpo->dpo_chunk_cnt++;
+	} else {
+		/* Reach the limit, Let's reuse the oldest (i.e. 1st) in list */
+		chunk = d_list_entry(dpo->dpo_chunk_list.next,
+				     struct daos_profile_chunk,
+				     dpc_chunk_list);
+		D_DEBUG(DB_TRACE, "Reuse the old profile buffer %p\n", chunk);
+	}
+
+	dpo->dpo_current_chunk = chunk;
+	return 0;
+}
+
+static int
+profile_op_init(struct daos_profile_op *dpo, int id, char *name)
+{
+	int rc;
+
+	dpo->dpo_op = id;
+	dpo->dpo_op_name = name;
+	D_INIT_LIST_HEAD(&dpo->dpo_chunk_list);
+	D_INIT_LIST_HEAD(&dpo->dpo_chunk_idle_list);
+	dpo->dpo_acc_cnt = 0;
+	dpo->dpo_acc_val = 0;
+	dpo->dpo_chunk_total_cnt = 0;
+	dpo->dpo_chunk_cnt = 0;
+
+	rc = profile_get_new_chunk(dpo);
+
+	return rc;
+}
+
+static void
+profile_chunk_next(struct daos_profile_op *dpo)
+{
+	struct daos_profile_chunk *chunk = dpo->dpo_current_chunk;
+
+	if (dpo->dpo_acc_cnt == 0)
+		return;
+
+	D_ASSERT(chunk != NULL);
+	D_ASSERT(chunk->dpc_chunk_offset <= chunk->dpc_chunk_size);
+
+	chunk->dpc_chunks[chunk->dpc_chunk_offset] =
+				dpo->dpo_acc_val / dpo->dpo_acc_cnt;
+	chunk->dpc_chunk_offset++;
+	dpo->dpo_acc_val = 0;
+	dpo->dpo_acc_cnt = 0;
+}
+
+static int
+profile_dump_chunk(struct daos_profile_op *dpo, FILE *file,
+		   struct daos_profile_chunk *dpc)
+{
+	char string[64] = { 0 };
```

Comment:
It might be helpful to #define a length and then use that instead of replicating '64' in the few places that used it.

Reply:
ok